pkg/integrations/v4/dm/id_provider.go

```diff
@@ -4,102 +4,151 @@
+	"context"
 	"fmt"
 	"github.com/newrelic/infrastructure-agent/pkg/backend/identityapi"
 	"github.com/newrelic/infrastructure-agent/pkg/entity"
 	"github.com/newrelic/infrastructure-agent/pkg/integrations/v4/protocol"
+	"sync"
 )
 
-type RegisteredEntitiesNameToID map[string]entity.ID
-type UnregisteredEntitiesNamed map[string]UnregisteredEntity
+type registeredEntitiesNameToID map[string]entity.ID
+type unregisteredEntitiesNamed map[string]unregisteredEntity
 type reason string
 
-type UnregisteredEntity struct {
+type unregisteredEntity struct {
 	Reason reason
 	Err    error
 	Entity protocol.Entity
 }
 
-type UnregisteredEntities []UnregisteredEntity
+type unregisteredEntityList []unregisteredEntity
+type unregisteredEntityListWithWait struct {
+	entities unregisteredEntityList
+	waitGroup *sync.WaitGroup
+}
+
+type entityListToRegisterWithWait struct {
+	entities []protocol.Entity
+	waitGroup *sync.WaitGroup
+}
 
 const reasonClientError = "Identity client error"
 const reasonEntityError = "Entity error"
+const reasonNotInCache = "Entity not cached yet"
 
-func newUnregisteredEntity(entity protocol.Entity, reason reason, err error) UnregisteredEntity {
-	return UnregisteredEntity{
+func newUnregisteredEntity(entity protocol.Entity, reason reason, err error) unregisteredEntity {
+	return unregisteredEntity{
 		Entity: entity,
 		Reason: reason,
 		Err:    err,
 	}
 }
 
 type idProviderInterface interface {
-	Entities(agentIdn entity.Identity, entities []protocol.Entity) (registeredEntities RegisteredEntitiesNameToID, unregisteredEntities UnregisteredEntities)
+	ResolveEntities(entities []protocol.Entity) (registeredEntitiesNameToID, unregisteredEntityListWithWait)
 }
 
-type idProvider struct {
+type cachedIdProvider struct {
 	client               identityapi.RegisterClient
-	cache                RegisteredEntitiesNameToID
-	unregisteredEntities UnregisteredEntitiesNamed
+	agentIdentity        func() entity.Identity
+	cache                registeredEntitiesNameToID
+	unregisteredEntities unregisteredEntitiesNamed
+	cacheMutex           *sync.Mutex
+	toRegisterQueue      chan entityListToRegisterWithWait
 }
 
-func NewIDProvider(client identityapi.RegisterClient) *idProvider {
-	cache := make(RegisteredEntitiesNameToID)
-	unregisteredEntities := make(UnregisteredEntitiesNamed)
-	return &idProvider{
+func NewCachedIDProvider(client identityapi.RegisterClient, agentIdentity func() entity.Identity, closeContext context.Context) *cachedIdProvider {
+	cache := make(registeredEntitiesNameToID)
+	unregisteredEntities := make(unregisteredEntitiesNamed)
+	cacheMutex := &sync.Mutex{}
+	provider := cachedIdProvider{
 		client:               client,
+		agentIdentity:        agentIdentity,
 		cache:                cache,
 		unregisteredEntities: unregisteredEntities,
+		cacheMutex:           cacheMutex,
+		toRegisterQueue:      make(chan entityListToRegisterWithWait, 10), // TODO adjust buffer
 	}
+
+	go provider.registerEntitiesWorker(closeContext)
```

Comment:
I would recommend to have a Run/Start method instead of starting goroutines from the "constructor"

Reply:
I don't see any problem here. Can you tell which problems can it bring?